CuDNN bindings rewrite (into ATen)

ezyang opened the pull request:
```
The executive summary is that this moves the torch/csrc/cudnn
library into ATen, adding a number of new cudnn_ methods to ATen
for batchnorm, convolution, affine grid generator and grid sampler.

ATen infra changes:

- TensorGeometry was moved to ATen
- TensorGeometry was modified to make its interface resemble that of
  Tensor; in particular, sizes is no longer a field, it's a method.
- AT_CUDA_ENABLED macro is set via ATen/Config.h header which is
  generated at cmake configure time.
  Fixes zdevito/ATen#168
- Change AT_CUDA_ENABLED macro to be a function macro, so that we
  error if it is not defined
- Introduce a new TensorArg class, which is a Tensor plus a little
  metadata.  This helps us give good error messages when checking
  dimensions/shapes of tensors.
  Fixes zdevito/ATen#169
- Also introduce a TensorGeometryArg class, for when you don't
  need the actual tensor data (which is most of the time.)
- Add ATen/Check.h, which contains a number of utility functions
  for testing shapes, types and devices of input tensors.  This
  will be particulary useful for native methods, which don't get
  code generated input testing code.  These functions take a
  'CheckedFrom' argument, at the moment just a string, which
  specifies some extra information about what function was
  doing the actual checking; this greatly improves error messages.
    - Many check functions take initializer lists, which let you
      test that all tensors have some property.  This API is
      peculiar, in that we IGNORE undefined tensors in this case.
      This is handled by filterDefined.
- Add AT_CUDNN_ENABLED macro
- CuDNN linking from ATen was improved; for example, we now actually
  add the CuDNN headers to our include path.
- Add some missing override specifiers to some methods
- We now actually build tests with CUDA functionality accessible
  (previously, AT_CUDA_ENABLED was not defined, meaning that
  the headers were missing all CUDA-only functionality.)
- Native functions now support giving explicit names to return
  outputs in yaml.  This makes it possible to hook into the NN
  autogenerated derivatives codepath using native functions.

CuDNN rewrite changes:

- torch/csrc/cudnn now uses ATen (rather than passing around
  THVoidTensor) and lives in ATen.  This lets us remove tensorPointer
  shenanigans.  The functions are exposed to ATen as native functions
  described in aten/src/ATen/cudnn/cuDNN.yaml
- ATen now builds and links against CuDNN when enabled.  The cmake
  package script was taken from Caffe2.
- Some header reorganization was done to help reduce dependencies
  on headers (this reorg is no longer used but I've kept it)
- Rename CHECK to CUDNN_CHECK
- Rip out old shape/type testing code in favor of modern ATen/Check.h
  interface using TensorArg.  In many cases, increase the robustness of
  the checking code.
- Change the inputs of the public facing functions, so that they can
  be bound by ATen
  - Delete THCState*; this is retrieved from the global ATen context
  - Delete cudnnHandle_t, this is retrieved from the global Handles.h
  - Delete cudnnDataType_t, this is retrieved from the Tensor type
  - Delete Convolution class, instead its constituent arguments are
    passed individually
- Change functions to return tensors, rather than take an appropriately
  sized output tensor as an input.
- Redo how transposed convolution / backward convolution is implemented
  (knock on effect of returning tensors).  Previously it was assumed
  that you would always pass an appropriately sized output tensor, but
  we don't want to do this anymore.  For backwards, we instead give
  the desired output tensor (input, really) size, because that is
  readily available.  For *transposed* convolution, however, we take
  output_padding, and otherwise do the shape calculation.
- Redo how legacy group convolution is implemented (knock on effect from
  porting cudnn to ATen.)  Previously, group convolution was implemented
  by manually constructing sizes and strides and then outputting
  appropriate, with macros switching between individual groups and
  all-at-once based on CuDNN version.  Now, the code looks exactly what
  you'd expect: there's a top-level wrapping function that supports
  group convolution no matter the version of CuDNN, and a low-level
  wrapper which supports only what CuDNN supports.  The top-level
  function conditions on CuDNN version, and invokes the low-level
  interface 1 or n times.
- There is now a debugging printer for tensor descriptors.
- Convolution struct is replaced with ConvolutionArgs, which is not
  part of the public API but is used internally to conveniently
  pass around all of the arguments needed for Convolution.
- Add some constexprs for well-known dimensions, reduce amount of
  magic numbers in code.
- Put 'deterministic' in to ConvParams.  Fixes #3659
- Lots more comments.
- Some pessimizations, in the name of code clarity:
  - The descriptors are initialized on every invocation of convolution
    forward/backward.  Previously, the descriptors were cached, so that
    you didn't have to initialize them again on backwards.  This is
    difficult to support in the ATen interface so I didn't support it.
  - Legacy group convolution initializes its workspace for *every* group
    it performs.  I did not feel motivated to fix this because the
    legacy codepath is already quite slow.
- Affine grid generator and grid sampler automatically call contiguous
  on their arguments as necessary.
- Batchnorm input checking is greatly beefed up, it now checks for
  the following input characteristics:
    - Definedness
    - GPU location
    - Type
    - Contiguity
    - Size

PyTorch binding code changes

- batchnorm now uses consistent var/data naming
- batchnorm and convolution make use of new ATen bindings
- Affine grid generator and grid sampler make use of ATen CuDNN
  bindings via derivatives.yaml.  This means I had to restructure
  the code a little, since the THNN bindings still go through
  a legacy Python class.
- I fixed some warnings:
  - s/friend class/friend struct/ on InterpreterStateImpl
  - Removed pessimizing move 'detached' in torch/csrc/autograd/variable.cpp
  - Removed unused pack_list on Scalar
```

ebetica commented:
@ezyang will there be an easily callable at::LSTM once this is done, which wraps CUDNN?

ezyang commented:
@ebetica Alas, LSTM is *not* handled by torch/csrc/cudnn, so not with this patch. However, making cuDNN LSTM callable from ATen is part of my mandate, so it would definitely make since as a follow up PR.

ezyang commented:
Latest commit makes it build but I have to rebase onto master (and reapply all changes since then on cudnn). I haven't actually tested this code at all yet :)

Biggest further substantive change on the last patch is, we do some more geometry calculation in convolution backwards, when previously it was done by saving sizes from forwards and then reusing them as needed.

ezyang commented:
More notes about changes:

* Some extra functions were added for handling conv versus conv_transpose. Hypothetically, you can just use conv where you need conv_transpose backward, and vice versa, but ACTUALLY (1) in the case of conv transpose and conv backward, the most convenient signatures for users in these two cases are different (one wants an `output_padding`, the other wants to just provide the actual expected output size and (2) you get better error messages this way, because the argument names are setup correctly (they would not be if you just called the "wrong" function.

* To assist in debugging check arguments failures, I beefed up `ATen/Check.h` to take an argument `CheckedFrom`, which is used to say what function we were checking the arguments for when it failed. This is especially helpful for backwards. This also got me thinking about call stacks; I wrote an issue at https://github.com/zdevito/ATen/issues/175

* Interestingly, the ported code *immediately* triggered the deterministic algorithm selection problem. Porting Natalias' fix solved this problem. This is probably because I fixed the benchmarking code to properly account for deterministic versus nondeterministic cases in parameter hashing, see https://github.com/pytorch/pytorch/issues/3659

* I added a CuDNN tensor descriptor printer for debugging purposes

This is a pretty big patchset. Should you trust it? Bugs that were not caught at compile time but caught at runtime:

* I accidentally swapped the input/output parameters for transposed convolution weight backwards. This caused an immediate `check` failure in the test suite. After fixing this I checked the original code for other invocations of `std::swap` but this particular case was the only one.

* I totally misunderstood the function of `output_padding` and thought that I should eliminate the asymmetry (actually, the asymmetry is necessary because output padding resolves the output size ambiguity in the case of transposed convolution.) My discoveries are now enshrined in code comments but they might be helpful in the user facing documentation as well.

* I accidentally messed up narrowing for groups by precomputing the tensor size per group once and using it for all values, even though actually the size varies between input and output. I fixed this by introducing a `narrowGroup` function which serves the same role as `tensorPointer` in the previous code.

* I misunderstood the API of `narrow` and thought that it took an end index in its second argument; it actually takes a length.

* I wrote some code for weight padding in batch norm incorrectly, and failed to expand it to 5D when the input was 5D. The current coding style is honestly a bit hazardous but I am not sure how to fix it (look for `wdesc{ weight->expand({1, weight->size(0)}), std::max(4, input->dim()) }`); the "near miss" for the other descriptors doesn't apply because descriptor construction with padding will take a max of the inputs; the problem is that weight isn't initially set to the correct dimension.

Affine grid sampler and generator are not yet wired up but they should be easy enough to do.

ezyang commented:
I'm going to have to eat my words. It looks like there's no way to bind ATen Tensor functions from Python automatically at the moment, so the only way I can expunge the old affine grid generator/sampler is by getting it all the way into full ATen gradient land at once.

ngimel commented on aten/src/ATen/cudnn/BatchNorm.cpp:
```diff
@@ -0,0 +1,168 @@
+  TensorDescriptor odesc{ *output, 4 };  // output descriptor
+  // TODO: For some reason, this was previously, inexplicably initialized from
+  // running_mean.  Check if there was a reason.
+  TensorDescriptor wdesc{ weight->expand({1, weight->size(0)}), std::max<int64_t>(4, input->dim()) };  // descriptor for weight, bias, running_mean, etc.
```
I don't know of a particular reason why it was initialized from running mean (@colesbury initially wrote these), but it should not matter, as all 6 (bias, weight, running_mean, running_var, save_mean, save_var) should share the same descriptor.

ngimel commented on aten/src/ATen/cudnn/BatchNorm.cpp:
```diff
@@ -0,0 +1,168 @@
+  cudnnSetStreamToCurrent();
+  checkSameType(c, {weight, bias, running_mean, running_var, save_mean, save_var});
+  // Check that if input is half, weight is float (but otherwise, they
+  // match)
```
There's still no check for half/float here?

ngimel commented on aten/src/ATen/cudnn/BatchNorm.cpp:
```diff
@@ -0,0 +1,168 @@
+  Constant one(dataType, 1);
+  Constant zero(dataType, 0);
+  if (training) {
+    checkContiguous(c, {input, bias, running_mean, running_var, save_mean, save_var});
```
is there a reason weight is not checked for being contiguous?

ezyang replied:
No, this is probabl an oversight.

ngimel commented on aten/src/ATen/cudnn/BatchNorm.cpp:
```diff
@@ -0,0 +1,168 @@
+  checkSameType(c, {input, grad_output});
+  checkSameType(c, {weight, running_mean, running_var, save_mean, save_var});
+  checkSameGPU(c, {input, grad_output, weight, running_mean, running_var, save_mean, save_var});
+  cudnnBatchNormMode_t mode;
```
Why are running_mean and running_var here? They are not used.

ngimel commented on aten/src/ATen/cudnn/BatchNorm.cpp:
```diff
@@ -0,0 +1,168 @@
+  TensorDescriptor idesc{ *input, 4 };  // input descriptor
+  TensorDescriptor odesc{ *grad_output, 4 };  // output descriptor
+  TensorDescriptor gdesc{ grad_input_t, 4 };  // grad_input descriptor
+  TensorDescriptor wdesc{ weight->expand({1, weight->size(0)}), std::max<int64_t>(4, input->dim()) };  // descriptor for weight, bias, running_mean, etc.
```
If input and grad_output are contiguous, technically it's enough to reuse the same descriptor for input, grad_output and grad_input.

ngimel commented on aten/src/ATen/cudnn/Descriptors.h:
```diff
@@ -0,0 +1,178 @@
+struct TensorDescriptor
+{
+  cudnnTensorDescriptor_t desc;
+  void* ptr;
```
I'm not sure data ptr should be part of TensorDescriptor. In cudnn model (however foolish) the same descriptor can be shared (and is shared, e.g. in batch norm) between several tensors, that's why it only has strides and sizes info, on purpose.

ezyang replied:
Hmm, this is an interesting point. I moved the data ptr into TensorDescriptor mostly to make the benchmark code look better after I deleted the `Convolution` struct, because otherwise I would have needed to add a big pile of extra parameters to thread the pointers through otherwise. I guess I can bring a `Convolution`-like struct to get rid of this situation.

ezyang commented:
This is fully ready for review, but I haven't handled @ngimel's comments yet.

apaszke commented on aten/src/ATen/cudnn/Utils.h:
```diff
@@ -0,0 +1,24 @@
+// TODO: Move this out to ATen proper?
+inline Tensor contiguousIfZeroInStrides(const Tensor& t) {
+  for (auto s : t.strides()) {
+    if (s == 0) return t.contiguous();
```
Whaaaaat? When is this function even used? It looks dangerous to me that you determine a tensor is contiguous based on a *single* stride value (having 0 stride means that this dim is contiguous, not whole tensor).

ezyang replied:
This code was lifted from some existing code in the affine grid generator bindings. My best explanation of what is going on here is this comment (which should be added here):

```
            # Sometimes grad_output is a scalar (like 1) expanded as a tensor.
            # cudnn requires a tensor that has non-zero strides.
```

apaszke replied:
I still don't get how does it explain anything. It's not a correct way to check that a tensor is an expanded scalar

ngimel replied:
I think what's going on is cudnn has a buggy check for tensor being contiguous (that is, it does not ignore stride for dimension that is equal to 1). And this function does not determine if tensor is contiguous, it makes a tensor that has a zero stride contiguous (hopefully setting strides for size=1 to cudnn't liking). If that's indeed the case, it should be commented better.

ezyang replied:
Commented.

apaszke commented on setup.py:
```diff
@@ -576,15 +571,6 @@ def check_file(f):
     include_dirs.insert(0, CUDNN_INCLUDE_DIR)
```
Shouldn't cuDNN be already linked to ATen?

ezyang replied:
Yes, but PyTorch also needs to link against CuDNN, because some codepaths, e.g., LSTM, go through this codepath.

apaszke commented on tools/autograd/templates/Functions.cpp:
```diff
@@ -417,7 +417,7 @@ Tensor softplus_double_backward(const Tensor & grad, const Tensor & input, Scala
 Tensor as_strided_backward(const Tensor & grad, TensorGeometry base, IntList sizes, IntList strides, int64_t storage_offset) {
   auto src = base.zeros_with_stride(grad.type());
-  src.as_strided(sizes, strides, storage_offset - base.storage_offset).copy_(grad);
+  src.as_strided(sizes, strides, storage_offset - base.storage_offset()).copy_(grad);
```
!! How did that even work before

ezyang replied:
This patchset contains a change to `storage_offset` in `TensorGeometry` to convert it into a method

apaszke commented on torch/backends/cudnn/__init__.py:
```diff
@@ -374,3 +379,21 @@ def descriptor_sequence(tensor, batch_sizes):
+        self.__dict__ = d
+    enabled = EnabledProp()
+
+sys.modules[__name__] = CudnnModule(sys.modules[__name__].__dict__)
```
this hack :smile:

ezyang replied:
Needs a comment haha

apaszke commented on torch/csrc/autograd/functions/convolution.cpp:
```diff
@@ -679,6 +663,7 @@ auto ConvBackwardBackward::apply(const variable_list& grad_grad_inputs) -> varia
       auto gOt = apply_fn<Transpose>(0, 1)(gO);
 
       // calculate output_padding
+      // TODO: figure out why this needs to be computed...
```
did you figure it out? :wink:

ezyang replied:
Nope

apaszke commented on torch/nn/_functions/vision.py:
```diff
@@ -9,6 +9,27 @@
+        if not cudnn.enabled:
+            raise RuntimeError("AffineGridGenerator needs CuDNN for "
+                               "processing CUDA inputs, but CuDNN is not enabled")
+        assert cudnn.is_acceptable(theta.data)
```
Don't use asserts for such checks. They can be disabled with command line flags

apaszke commented on aten/src/ATen/cudnn/Conv.cpp:
```diff
@@ -0,0 +1,1164 @@
+constexpr int output_batch_size_dim = 0;  // also grad_output
+constexpr int output_channels_dim = 1;
+constexpr int weight_output_channels_dim = 0;
+constexpr int weight_input_channels_dim = 1;
```
:+1: :+1: :+1:

apaszke commented on aten/src/ATen/cudnn/Conv.cpp:
```diff
@@ -0,0 +1,1164 @@
+    IntList padding, IntList stride, IntList dilation, int groups
+) {
+  // ASSERT(input_size.size() > 2)
+  // ASSERT(input_size.size() == weight_size.size())
```
Either uncomment or remove those

apaszke commented on aten/src/ATen/cudnn/Conv.cpp:
```diff
@@ -0,0 +1,1164 @@
+}
+
+// TODO: Move this into the standard library, with a better name?
+Tensor narrowGroup(const Tensor& t, int dim, int group_idx, int groups) {
```
BTW this narrowing could have been implemented as a single `split` at the beginning of each function

ezyang replied:
I... suppose? IIRC, doesn't split repeatedly call narrow internally anyway? So it doesn't seem to make a difference

apaszke commented on aten/src/ATen/cudnn/Conv.cpp:
```diff
@@ -0,0 +1,1164 @@
+    throw std::runtime_error(ss.str());
+  }
+
+  auto num_negative_values = std::count_if(args.begin(), args.end(), [](int x){return x < 0;});
```
nit: `std::any` would be sufficient

apaszke commented on aten/src/ATen/cudnn/Conv.cpp:
```diff
@@ -0,0 +1,1164 @@
+  if (num_negative_values > 0){
+    std::stringstream ss;
+    ss << arg_name << " should be greater than zero but got (";
+    std::copy(args.begin(), args.end() - 1, std::ostream_iterator<int>(ss,", "));
```
woooow, that's an awesome shortcut to print a vector! The only thing I don't understand is why you do `args.end() - 1`. `std::copy` works on the interval `[first, last)`

ezyang replied:
Not my shortcut, this is from the old code :) I dunno how this trick works.

apaszke commented on aten/src/ATen/cudnn/Conv.cpp:
```diff
@@ -0,0 +1,1164 @@
+    }
+  }
+  // TODO: check that output->size() matches output_sizes
+  // TODO: check that weight matches output->sizes()
```
These should be implemented before merge

apaszke commented on aten/src/ATen/cudnn/Conv.cpp:
```diff
@@ -0,0 +1,1164 @@
+    params->weight_size[i] = (int) weight.size(i);
+  }
+  // ASSERT(padding.size() == stride.size())
+  // ASSERT(padding.size() == dilation.size())
```
Uncomment or delete

apaszke commented on aten/src/ATen/cudnn/Conv.cpp:
```diff
@@ -0,0 +1,1164 @@
+BenchmarkCache<cudnnConvolutionBwdFilterAlgo_t> bwd_filter_algos;
+
+// TODO: Stop manually allocating CUDA memory; go through ATen
+// (ATen should provide an API for this!)
```
ATen provides an API for this - allocate a CUDA byte tensor/storage

ezyang replied:
Won't a storage allocation come with a bunch of TH metadata that I don't actually need here?

apaszke replied:
It would but what's the difference? Benchmarking is the slow path anyway

apaszke reviewed (commented):
The code looks really neat, I only have some very minor comments. However, I haven't thoroughly reviewed Conv.cpp (mostly skimmed over it)

ezyang commented on aten/src/ATen/cudnn/ConvBenchmark.h:
```diff
@@ -0,0 +1,19 @@
+
+// Hashing mach
+
+}}
```
Oops, should be deleted.

ezyang commented on aten/src/ATen/Check.cpp:
```diff
@@ -0,0 +1,123 @@
+std::ostream& operator<<(std::ostream & out, TensorGeometryArg t) {
+  if (t.pos == 0) {
+    // 0 is distinguished; it usually indicates 'self' or the return
+    // tensor
```
I don't understand this comment, because the logic here is about controlling whether or not we say 'argument #1' or not; we always print the name.

ezyang commented on aten/src/ATen/Check.h:
```diff
@@ -0,0 +1,53 @@
+struct TensorArg {
+  Tensor tensor;
+  const char* name;
+  int pos; // 1-indexed
```
I'm not sure what else you want to do ;) Given `f(x, y)`, is `x` argument 0 or 1? PyTorch's local conventions seem to be to describe it as argument 1.

ezyang commented on aten/src/ATen/Check.h:
```diff
@@ -0,0 +1,53 @@
+  const Tensor& operator*() const { return tensor; }
+};
+
+struct TensorGeometryArg {
```
I don't feel like it :P

ezyang commented on aten/src/ATen/Check.h:
```diff
@@ -0,0 +1,53 @@
+  const TensorGeometry& operator*() const { return tensor; }
+};
+
+using CheckedFrom = std::string;
```
Yes. Sure, we can make it a char* for now. I was thinking maybe we'd like to make it a little more complicated in the future, possibly.

ezyang commented:
OMG build is finally green.

Latest changelog:

- I expanded the native yaml parser *just* enough to make it possible for me to explicitly specify the names of multiple return outputs. This allows me to use the NN autogenerated wrapper codepath, which assumes these names are coherent.

- BatchNorm input checking is now greatly expanded; I basically went through definedness, GPU, type, contiguity, size, and made sure I had checking in all of these cases. I should also expand this checking to other functions but I haven't gotten around to it yet.

- As suggested by the reviews, descriptors have lost the data pointer, and there is a new ConvolutionArgs struct which contains everything we need to do benchmarking. BTW, I noticed we don't seem to actually exercise the benchmark code in our test suite.

- I messed up the no CUDA build from deleting some necessary preprocessor ifdefs. These have been added back.

- I added a few misc warning squishes.

- I removed the "propagate cudnn enabled to ATen" patch, as it was failing CI in a hard to understand way. I'll PR it separately.

I'll squash the patchset before merge.

ezyang committed:
[BREAKS BUILD] Move torch/csrc/cudnn to aten/src/ATen/cudnn

This is just to make it easier to read the diff in the
subsequent commit.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Comprehensive rewrite of Torch CuDNN bindings / a bit of ATen infra

The executive summary is that this moves the torch/csrc/cudnn
library into ATen, adding a number of new cudnn_ methods to ATen
for batchnorm, convolution, affine grid generator and grid sampler.

ATen infra changes:

- TensorGeometry was moved to ATen
- TensorGeometry was modified to make its interface resemble that of
  Tensor; in particular, sizes is no longer a field, it's a method.
- AT_CUDA_ENABLED macro is set via ATen/Config.h header which is
  generated at cmake configure time.
  Fixes https://github.com/zdevito/ATen/issues/168
- Change AT_CUDA_ENABLED macro to be a function macro, so that we
  error if it is not defined
- Introduce a new TensorArg class, which is a Tensor plus a little
  metadata.  This helps us give good error messages when checking
  dimensions/shapes of tensors.
  Fixes https://github.com/zdevito/ATen/issues/169
- Also introduce a TensorGeometryArg class, for when you don't
  need the actual tensor data (which is most of the time.)
- Add ATen/Check.h, which contains a number of utility functions
  for testing shapes, types and devices of input tensors.  This
  will be particulary useful for native methods, which don't get
  code generated input testing code.  These functions take a
  'CheckedFrom' argument, at the moment just a string, which
  specifies some extra information about what function was
  doing the actual checking; this greatly improves error messages.
    - Many check functions take initializer lists, which let you
      test that all tensors have some property.  This API is
      peculiar, in that we IGNORE undefined tensors in this case.
      This is handled by filterDefined.
- Add AT_CUDNN_ENABLED macro
- CuDNN linking from ATen was improved; for example, we now actually
  add the CuDNN headers to our include path.
- Add some missing override specifiers to some methods
- We now actually build tests with CUDA functionality accessible
  (previously, AT_CUDA_ENABLED was not defined, meaning that
  the headers were missing all CUDA-only functionality.)
- Native functions now support giving explicit names to return
  outputs in yaml.  This makes it possible to hook into the NN
  autogenerated derivatives codepath using native functions.

CuDNN rewrite changes:

- torch/csrc/cudnn now uses ATen (rather than passing around
  THVoidTensor) and lives in ATen.  This lets us remove tensorPointer
  shenanigans.  The functions are exposed to ATen as native functions
  described in aten/src/ATen/cudnn/cuDNN.yaml
- ATen now builds and links against CuDNN when enabled.  The cmake
  package script was taken from Caffe2.
- Some header reorganization was done to help reduce dependencies
  on headers (this reorg is no longer used but I've kept it)
- Rename CHECK to CUDNN_CHECK
- Rip out old shape/type testing code in favor of modern ATen/Check.h
  interface using TensorArg.  In many cases, increase the robustness of
  the checking code.
- Change the inputs of the public facing functions, so that they can
  be bound by ATen
  - Delete THCState*; this is retrieved from the global ATen context
  - Delete cudnnHandle_t, this is retrieved from the global Handles.h
  - Delete cudnnDataType_t, this is retrieved from the Tensor type
  - Delete Convolution class, instead its constituent arguments are
    passed individually
- Change functions to return tensors, rather than take an appropriately
  sized output tensor as an input.
- Redo how transposed convolution / backward convolution is implemented
  (knock on effect of returning tensors).  Previously it was assumed
  that you would always pass an appropriately sized output tensor, but
  we don't want to do this anymore.  For backwards, we instead give
  the desired output tensor (input, really) size, because that is
  readily available.  For *transposed* convolution, however, we take
  output_padding, and otherwise do the shape calculation.
- Redo how legacy group convolution is implemented (knock on effect from
  porting cudnn to ATen.)  Previously, group convolution was implemented
  by manually constructing sizes and strides and then outputting
  appropriate, with macros switching between individual groups and
  all-at-once based on CuDNN version.  Now, the code looks exactly what
  you'd expect: there's a top-level wrapping function that supports
  group convolution no matter the version of CuDNN, and a low-level
  wrapper which supports only what CuDNN supports.  The top-level
  function conditions on CuDNN version, and invokes the low-level
  interface 1 or n times.
- There is now a debugging printer for tensor descriptors.
- Convolution struct is replaced with ConvolutionArgs, which is not
  part of the public API but is used internally to conveniently
  pass around all of the arguments needed for Convolution.
- Add some constexprs for well-known dimensions, reduce amount of
  magic numbers in code.
- Put 'deterministic' in to ConvParams.  Fixes #3659
- Lots more comments.
- Some pessimizations, in the name of code clarity:
  - The descriptors are initialized on every invocation of convolution
    forward/backward.  Previously, the descriptors were cached, so that
    you didn't have to initialize them again on backwards.  This is
    difficult to support in the ATen interface so I didn't support it.
  - Legacy group convolution initializes its workspace for *every* group
    it performs.  I did not feel motivated to fix this because the
    legacy codepath is already quite slow.
- Affine grid generator and grid sampler automatically call contiguous
  on their arguments as necessary.
- Batchnorm input checking is greatly beefed up, it now checks for
  the following input characteristics:
    - Definedness
    - GPU location
    - Type
    - Contiguity
    - Size

PyTorch binding code changes

- batchnorm now uses consistent var/data naming
- batchnorm and convolution make use of new ATen bindings
- Affine grid generator and grid sampler make use of ATen CuDNN
  bindings via derivatives.yaml.  This means I had to restructure
  the code a little, since the THNN bindings still go through
  a legacy Python class.
- I fixed some warnings:
  - s/friend class/friend struct/ on InterpreterStateImpl
  - Removed pessimizing move 'detached' in torch/csrc/autograd/variable.cpp
  - Removed unused pack_list on Scalar

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

GCC 4.8 buildfix

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Add TensorGeometry to ATen.h

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

CUDNN_CHECK

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Update TODO comment

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Delete return in cudnn_grid_sampler

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

s/cudnnSetStreamToCurrent/setCuDNNStreamToCurrent/g

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Don't allocate a new vector when filtering defined.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Remove Check overloads, convert to pass references.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Some more microbenchmarking.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

zdevito commented on aten/src/ATen/native_parse.py:
```diff
@@ -54,11 +54,16 @@ def run(paths):
-            func_decl, return_type = [x.strip() for x in func['func'].split('->')]
+            if '->' in func['func']:
+                func_decl, return_type = [x.strip() for x in func['func'].split('->')]
+                return_type = sanitize_types(return_type)
```
sanitize_types would be a good place to make everything `type: name:`. Also it would be easy to support named return types here as well.

zdevito reviewed (commented):
I reviewed mostly how cudnn integrated into ATen and did not look carefully at the  checks in cudnn methods themselves. I think this is really cool. I have just a few general comments/questions:

* There seems to be a lot of unintentional or at least surprising vector allocations in the checking code. Functions that call 5 checking functions might do 20 malloc/free pairs and 20 retain/releases just for checking. It might help to distinguish between TensorGeometries that are long term and need copies and TensorGeometries that are just a view of existing stuff and can use ArrayLists.

* What remains to eliminate cudnn cpp functions and just go through aten directly for cudnn operators?

* Do you have any idea of the launch performance before/after this pathway to catch potential performance regressions?

ngimel commented on aten/src/ATen/cudnn/Conv.cpp:
```diff
@@ -0,0 +1,1164 @@
+
+// TODO: Stop manually allocating CUDA memory; go through ATen
+// (ATen should provide an API for this!)
+struct Workspace {
```
is ```at::Tensor``` of type ```kByte``` guaranteed to give nicely aligned memory? Aligning memory to byte won't work, by and large.

colesbury replied:
The data type doesn't affect alignment. For CUDA memory we either get something from `cudaMalloc` (which is "suitably aligned") or a split cached block (which 512-byte aligned).

ezyang replied:
I'll update the TODO comment with this implementation guidance, but I am too lazy to actually do this.

apaszke commented on aten/src/ATen/cudnn/Conv.cpp:
```diff
@@ -0,0 +1,1164 @@
+    };
+    // NOTE: - 1 because ALGO_WINOGRAD is not implemented
+    static constexpr int num_algos = CUDNN_CONVOLUTION_BWD_FILTER_ALGO_COUNT - 1;
+    static_assert(sizeof(algos) / sizeof(algos[0]) == num_algos,
```
Yeah, I added this some time ago, and that's the whole point of this change. Once new cuDNN comes out, we can be sure we won't miss any algorithms.

ezyang commented on aten/src/ATen/cudnn/cuDNN.yaml:
```diff
@@ -0,0 +1,71 @@
+  variants: function
+
+# NB: input is special cased in a way I don't quite understand
+- func: cudnn_grid_sampler(Tensor self, Tensor grid) -> Tensor
```
Yes, the problem is there is no syntax for named return types, and NN autogen relies on the return name. Not terribly motivated to fix this if we're axing NN autogen though.

Also, yes I'll delete the return in `func`.

ezyang commented:
> There seems to be a lot of unintentional or at least surprising vector allocations in the checking code. Functions that call 5 checking functions might do 20 malloc/free pairs and 20 retain/releases just for checking. It might help to distinguish between TensorGeometries that are long term and need copies and TensorGeometries that are just a view of existing stuff and can use ArrayLists.

I wrote this code assuming that small vector allocation was cheap. I suppose it's not, so I better clean it up...

> What remains to eliminate cudnn cpp functions and just go through aten directly for cudnn operators?

- [ ] Teach ATen to link with NNPack
- [ ] Port CppFunction bodies to ATen and put them through derivatives
- [ ] Port cudnn double backwards in Python to ATen
- [ ] Propagate cudnn-enabled to ATen context (patch constructed)

> Do you have any idea of the launch performance before/after this pathway to catch potential performance regressions?

No, I guess I better go test this.

ezyang commented on aten/src/ATen/Check.cpp:
```diff
@@ -0,0 +1,213 @@
+namespace at {
+
+namespace {
+  std::vector<TensorArg> filterDefined(ArrayRef<TensorArg> ts) {
```
Oh how I wish C++ had stream fusion haha. I'll fix this.

ezyang commented on aten/src/ATen/TensorGeometry.h:
```diff
@@ -0,0 +1,39 @@
+    return r;
+  }
+
+  std::vector<int64_t> sizes_;
```
I went ahead and changed all the check functions to accept `TensorArg` and `TensorGeometryArg` by reference.

ebetica commented:
@ezyang is CUDNN for Conv enabled if we don't care about fancy stuff like double backward? If so, what are the code paths to enable or automatically detect it? If I just call at::conv2d will it automagically work?

ezyang commented:
Not yet, I need to wire this up.
